code/modules/integrated_electronics/subtypes/insert_slot.dm

```diff
@@ -0,0 +1,74 @@
+//Insert slots allow items to be inserted into assemblies from outside
+//These items can then be used by other components
+/obj/item/integrated_circuit/insert_slot
+	category_text = "Insert slot"
+	var/capacity = 0
+	var/list/allowed_types = list()
+	var/list/items_contained = list()
+	activators = list("eject contents" = IC_PINTYPE_PULSE_IN)
+	outputs = list("has item" = IC_PINTYPE_BOOLEAN)
+	power_draw_per_use = 1 
+	w_class = ITEMSIZE_NORMAL
+	size = 5
+	complexity = 1
+
+//call this function from components that want to get items from this component
+//set remove to FALSE if you dont want the item removed from the component and just want a reference to it 
+//(e.g. for beakers)
+/obj/item/integrated_circuit/insert_slot/proc/get_item(var/remove = FALSE)
+	if(items_contained.len > 0)
+		var/itemToReturn = items_contained[1]
+		if(remove)
+			items_contained -= itemToReturn
+			if(items_contained.len <= 0)
+				set_pin_data(IC_OUTPUT, 1, FALSE)
+				push_data()
+		return itemToReturn
+	return null
+
+/obj/item/integrated_circuit/insert_slot/do_work()
+	for(var/obj/item/O in items_contained)
+		O.forceMove(get_turf(src))
+		visible_message("<span class='notice'>\The [src] drops [O] on the ground.</span>")
+		items_contained -= O
+	set_pin_data(IC_OUTPUT, 1, FALSE)
+	push_data()
+	return TRUE
+
+/obj/item/integrated_circuit/insert_slot/proc/insert(var/obj/item/O, var/mob/user)
+	for(var/next_type in allowed_types)
+		if(istype(O, next_type))
+			if(items_contained.len >= capacity)
+				to_chat(user, "<span class='warning'>\The [src] is too full to add [O].</span>")
+				return FALSE
+			items_contained += O
+			user.drop_item(O)
+			O.forceMove(src)
+			to_chat(user, "<span class='notice'>You add [O] to \the [src].</span>")
+			set_pin_data(IC_OUTPUT, 1, TRUE)
+			return TRUE
+		else
+			to_chat(user, "<span class='notice'>You can not add [O.type] to \the [src] because it is not \a [allowed_types[1]]</span>")
```

Comment:
If you have two types in allowed_types (/obj/item/a, /obj/item/b) and you try to insert /obj/item/b wouldnt that print that print this error message for /obj/item/a ?
(Since you are looping over all allowed types and check if the object someone tires to insert is next_type and print a error message if it is not ?)
`for(var/next_type in allowed_types)`